### PR TITLE
dm-master: leader election

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

add leader election supported for DM-master based on embed etcd

### What is changed and how it works?

do leader election based on embed etcd

some code learn from https://github.com/pingcap/tidb/blob/v3.0.5/owner/manager.go.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to be included in the release note

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@WangXiangUSTC , @GregoryIan or @amyangfei   PTAL

**Comment by csuzhangxc** [comment]
I update the methods for `Election` after added `PTAL` label. It's ready for review again now.

File: tests/others_integration.txt [hunk 1]
```diff
@@ -1,7 +1 @@
-all_mode
-sequence_sharding
-sequence_safe_mode
-relay_interrupt
 start_task
-initial_unit
-http_apis
```
Comment (csuzhangxc) [inline tests/others_integration.txt]: I have updated the CI to run these cases separately

File: dm/master/election.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+
+	"github.com/pingcap/dm/pkg/log"
+)
+
+func (s *Server) electionNotify(ctx context.Context) {
```
Comment (WangXiangUSTC) [inline dm/master/election.go]: why need this function, seems `toBeLeader` and `retireLeader` already print the log

File: pkg/election/election.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+func (e *Election) retireLeader() {
+	e.isLeader.Set(false)
+	select {
+	case e.leaderCh <- false:
+	default:
+	}
+	e.l.Info("retire from the leader", zap.String("member", e.id))
+}
+
+func (e *Election) watchLeader(ctx context.Context, session *concurrency.Session, key string) {
```
Comment (WangXiangUSTC) [inline pkg/election/election.go]: the key can only be e.key? if so don't need put it in arguments.

File: dm/master/server.go [hunk 4]
```diff
@@ -151,14 +166,30 @@ func (s *Server) Start(ctx context.Context) (err error) {
...
+	// NOTE: we only use the local member's address now, but we can use all endpoints of the cluster if needed.
+	s.etcdClient, err = etcdutil.CreateClient([]string{s.cfg.MasterAddr})
+	if err != nil {
+		return
+	}
+
 	s.closed.Set(false) // the server started now.
 
+	// start leader election
+	s.election = election.NewElection(ctx, s.etcdClient, electionTTL, electionKey, s.cfg.Name)
```
Comment (WangXiangUSTC) [inline dm/master/server.go]: do we need to handle `election`'s error from `election.ErrorNotify()`

File: pkg/election/election.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+				return
+			}
+			if resp.Canceled {
+				e.l.Info("watch canceled")
+				return
+			}
+
+			for _, ev := range resp.Events {
+				if ev.Type == mvccpb.DELETE {
+					e.l.Info("fail to watch, the leader is deleted", zap.ByteString("key", ev.Kv.Key))
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/election/election.go]: user may force to trigger a new campaign by deleting the key with _etcdctl_ (or _dmctl_ in the future); in the normal case, before electionTTL overdue, `Session` will `keep alive for the lifetime of a client`.

see:
- https://github.com/etcd-io/etcd/blob/v3.4.3/clientv3/concurrency/session.go#L26
- https://github.com/etcd-io/etcd/blob/v3.4.3/clientv3/concurrency/session.go#L63

File: dm/master/election.go (hunk 2 above)
Comment (csuzhangxc) [reply dm/master/election.go]: `toBeLeader` and `retireLeader` only log the leadership of the current member.
this function is used to show the usage of `Election` (and also log the leader name even if it's not the current member), do you think should we need to remove it?

File: pkg/election/election.go (hunk 3 above)
Comment (csuzhangxc) [reply pkg/election/election.go]: `e.key` is the prefix of this `key`, like `e.key == /dm-master/leader`, but `key == /dm-master/leader/105a6e8c8aee0806` (with the `lease` as the suffix), and this key can change for every campaign loop.

File: dm/master/server.go (hunk 4 above)
Comment (csuzhangxc) [reply dm/master/server.go]: I update `Election` to let it more like which one in TiDB's owner pkg in https://github.com/pingcap/dm/pull/367/commits/80ae3d23a14157b78fdc4b7ac2a16e424ad18fef and https://github.com/pingcap/dm/pull/367/commits/5c4c8204aeb8eca110439574ea63596c6901752b.

1. Detect any potential error (for unrecoverable error) in `NewElection`, and do not start `Server` if an error occurred.
2. Retry for errors (for recoverable error) until the context is done after entering the campaign loop.
3. receive errors from `election.ErrorNotify()` in _server/election.go_, but do no meaningful things now.

**Comment by csuzhangxc** [comment]
@WangXiangUSTC PTAL again

File: dm/master/election.go (hunk 2 above)
Comment (csuzhangxc) [reply dm/master/election.go]: one more thing, `pkg/election` is a pkg, but this function is in the application. Maybe we need to add some metrics for the leadership later, and do that in this function is better.

File: pkg/election/election.go (hunk 3 above)
Comment (WangXiangUSTC) [reply pkg/election/election.go]: ok

File: pkg/election/election.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/election/election.go]: how about adding a comment in which case will get delete event

File: dm/master/election.go (hunk 2 above)
Comment (WangXiangUSTC) [reply dm/master/election.go]: if master needs to do something(for example: meet error in the election, master exit) according to the election's result I think putting here is better. 
it is up to you

File: pkg/election/election.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+		if i > 0 {
+			select {
+			case e.ech <- terror.ErrElectionCampaignFail.Delegate(err, "create a new session"):
+			default:
+			}
+
+			select {
+			case <-time.After(newSessionRetryInterval):
+			case <-ctx.Done():
+				break
```
Comment (WangXiangUSTC) [inline pkg/election/election.go]: do you means break the for loop, this break only break the `select`
https://stackoverflow.com/questions/11104085/in-go-does-a-break-statement-break-from-a-switch-select

File: pkg/election/election.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/election/election.go]: added in https://github.com/pingcap/dm/pull/367/commits/b99ae4653c5039cfb5070e5b0d62d6bdad77599a.

File: dm/master/election.go (hunk 2 above)
Comment (csuzhangxc) [reply dm/master/election.go]: I chose to remove the log in `pkg/election` in https://github.com/pingcap/dm/pull/367/commits/b99ae4653c5039cfb5070e5b0d62d6bdad77599a.

File: pkg/election/election.go (hunk 6 above)
Comment (csuzhangxc) [reply pkg/election/election.go]: good catch!, fixed in https://github.com/pingcap/dm/pull/367/commits/b99ae4653c5039cfb5070e5b0d62d6bdad77599a.

**Comment by WangXiangUSTC** [comment]
LGTM

File: dm/master/election.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+		case leader := <-s.election.LeaderNotify():
+			// output the leader info.
+			if leader {
+				log.L().Info("current member become the leader", zap.String("current member", s.cfg.Name))
+			} else {
+				_, leaderID, err2 := s.election.LeaderInfo(ctx)
+				if err2 == nil {
+					log.L().Info("current member retire from the leader", zap.String("leader", leaderID), zap.String("current member", s.cfg.Name))
+				} else {
+					log.L().Error("get leader info", zap.Error(err2))
```
Comment (amyangfei) [inline dm/master/election.go]: Is `Warn` level better

File: pkg/election/election.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+			case <-ctx.Done():
+				break forLoop
+			}
+		}
+
+		// add more options if needed.
+		// NOTE: I think use the client's context is better than something like `concurrency.WithContext(ctx)`,
+		// so we can close the session when the client is still valid.
+		session, err = concurrency.NewSession(e.cli, concurrency.WithTTL(e.sessionTTL))
+		if err == nil || errors.Cause(err) == ctx.Err() {
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/election/election.go]: updated in https://github.com/pingcap/dm/pull/367/commits/708c2c1288d3bd3a379399d6e22f76572488b9fb.

File: dm/master/election.go (hunk 7 above)
Comment (csuzhangxc) [reply dm/master/election.go]: updated in https://github.com/pingcap/dm/pull/367/commits/708c2c1288d3bd3a379399d6e22f76572488b9fb.

**Pull request merged**